### ISSUE TITLE
Library lacks direct square free querying functionality

### ISSUE THREAD

**Issue by GregoryMorse** [sympy/sympy#27271]
I would propose as it would enhance the library to add to `sympy.ntheory` possibly under `factor` an `is_square_free(t=2)` function that more efficiently computes the t-th square-freeness of a number.

Currently the following idioms can do this but in a non-efficient manner as this can be short-circuited:
1. `all(x<t for _, x in sympy.ntheory.factorint(n).items())`
2. `sympy.ntheory.factor_.core(n, t)==n`

**Comment by RushabhMehta2005** [comment]
Hello, this is the rough code for the is_square_free function. I am confused about its expected behaviour on invalid inputs. Do I raise a ValueError for n <= 0 and t < 2? 

```python
from sympy.ntheory import factorint

def is_square_free(n, t=2):
    r"""
    Checks if a given integer is t-th square-free.

    A positive integer `n` is said to be t-th square-free if no prime factor 
    of `n` appears with an exponent greater than or equal to `t`. By default, 
    this function checks for square-freeness (t=2), meaning no prime factor 
    appears with an exponent of 2 or greater.

    Parameters
    ==========

    n : int
        A positive integer to be checked for t-th square-freeness.
    t : int, optional
        The value of `t` for t-th square-freeness. Defaults to 2.

    Returns
    =======

    bool
        True if `n` is t-th square-free, False otherwise.

    Examples
    ========

    >>> from sympy.ntheory.factor_ import is_square_free
    >>> is_square_free(10)  # 10 = 2^1 * 5^1 (t=2)
    True
    >>> is_square_free(12)  # 12 = 2^2 * 3^1 (t=2)
    False
    >>> is_square_free(18, t=3)  # 18 = 2^1 * 3^2 (t=3)
    True
    >>> is_square_free(16, t=3)  # 16 = 2^4 (t=3)
    False

    Notes
    =====

    - `n` must be a positive integer; otherwise, a `ValueError` is raised.
    - If `t < 2`, a `ValueError` is raised as t-th square-freeness is only 
      meaningful for `t >= 2`.
    - For `n = 1`, the function always returns True, as 1 is t-th square-free
      for all values of `t`.
    """

    if n <= 0:
        raise ValueError("Only positive integers can be checked for t-th square-freeness.")
    if n == 1:
        return True
    if t < 2:
        raise ValueError("t must be at least 2 for meaningful t-th square-freeness.")

    factors = factorint(n)

    # Check if any prime factor has an exponent >= t
    for exponent in factors.values():
        if exponent >= t:
            return False

    return True

# Usage
print(is_square_free(10, 2))  # True, 10 = 2^1 * 5^1
print(is_square_free(12, 2))  # False, 12 = 2^2 * 3^1 
print(is_square_free(72, 3))  # False, 72 = 2^3 * 3^2
print(is_square_free(72, 4))  # True, 72 = 2^3 * 3^2
print(is_square_free(18, 3))  # True, 18 = 2 * 3^2
print(is_square_free(16, 3))  # False, 16 = 2^4

**Comment by oscarbenjamin** [comment]
> t-th square-freeness of a number.

Is there a better name for this?

**Comment by haru-44** [comment]
> > t-th square-freeness of a number.
> 
> Is there a better name for this?

The term "k-th power free integers" seems to be appropriate.

https://mathworld.wolfram.com/Powerfree.html

**Comment by GregoryMorse** [comment]
The code is not efficient though as `factorint` does not return an iterator for short-circuit evaluation.  All of the hard work will already have been done.

Yes `is_square_free(n)` can call `is_kth_power_free(n, t=2)` and `square_free_range(a, b)` calls `kth_power_free_range(a, b, t=2)`.

**Comment by RushabhMehta2005** [comment]
Right, since factorint becomes the bottleneck and we want to return False immediately after first occurence of a prime with exponent >= t, we need a way to get the exponents step by step. How might we approach this? If there was something close to the factorint that yielded a value then it would be easy to solve.

**Comment by RushabhMehta2005** [comment]
> Yes `is_square_free(n)` can call `is_kth_power_free(n, t=2)` and `square_free_range(a, b)` calls `kth_power_free_range(a, b, t=2)`.

I am confused about the purpose of the function `is_kth_power_free(n, t=2)` here.

**Comment by GregoryMorse** [comment]
The square free versions are just light wrappers around the general kth power free one so just a forwarding call for t==2 as square is a special case worth explicit naming.  I checked and core() is implemented just as you implement with factorint which should be rewritten to return some sort of iterator as it would improve a lot of functions.

For now the code above if naming fixed is fine, as factorint is too complex to be worth duplicating.  factorint should call and accumulate _factorint which returns a generator.  Whether it should generate p one by one is probably not as efficient as generating tuples (p,e)

**Comment by RushabhMehta2005** [comment]
Alright I'll change the naming of this function. I'll also work on the square_free_range function and then make a PR. I'll add tests as well.

**Comment by vprayag2005** [comment]
here is the rough code of is_kth_power_free 
any suggestion and improvement
from sympy.ntheory.factor_  import primefactors
def is_kth_power_free(n,t):
    if n <= 0:
        raise ValueError("The number must be a positive integer to check for t-th power-freeness.")
    if n == 1:
        return True
    if t < 2:
        raise ValueError("The value of t must be 2 or greater to check for t-th power-freeness.")
    primefactors_list = primefactors(n)
    for i in primefactors_list:
        if n % (i ** t) == 0:
            return False
    return True

**Comment by oscarbenjamin** [comment]
I'm going to quote what I said in the associated PR that tried to add a function like the one just shown https://github.com/sympy/sympy/pull/27286#issuecomment-2491126760:

We don't need 300 lines of code, docs etc just to make trivial wrappers around an inefficient oneliner:
```python
all(e < k for e in factorint(n).values())
```
I suggest not to include this. I don't think that these functions should ever be used in sympy itself and I don't think that others should use them either.

This would be the way to make something that is potentially more useful than just calling factorint:

> Ideally, it would be best if factorint could use yield to report each prime factor as it is found, but implementing such a modification would be challenging.

Maybe it would be challenging to modify factorint like that but I would say that then it is not worth adding anything here. A note could be added to the factorint docs for how to check if a number is square-free.
